concent_api/core/payments/bankster.py

```diff
@@ -384,23 +381,68 @@ def sum_subtask_price(subtask_results_accepted_list: List[SubtaskResultsAccepted
 
 def get_provider_payment_info(
     list_of_forced_payments: List[ForcedPaymentEvent],
-    list_of_payments: List[BatchTransferEvent],
+    already_satisfied_claims: QuerySet,
     subtask_results_accepted_list: List[SubtaskResultsAccepted],
 ) -> Tuple[int, int]:
-    assert isinstance(list_of_payments, list)
+    assert isinstance(already_satisfied_claims, QuerySet)
     assert isinstance(list_of_forced_payments, list)
     assert isinstance(subtask_results_accepted_list, list)
 
-    force_payments_price = sum_payments(list_of_forced_payments)
-    payments_price = sum_payments(list_of_payments)
+    forced_satisfied_payments_price = sum_payments(list_of_forced_payments)
+    satisfied_payments_price = already_satisfied_claims.aggregate(
+        sum_of_already_satisfied_claims=Coalesce(Sum('amount'), 0)
+    )['sum_of_already_satisfied_claims']
     subtasks_price = sum_subtask_price(subtask_results_accepted_list)
 
-    amount_paid = payments_price + force_payments_price
+    amount_paid = satisfied_payments_price + forced_satisfied_payments_price
     amount_pending = subtasks_price - amount_paid
 
     return (amount_paid, amount_pending)
 
 
+def remove_duplicated_transactions_from_deposit_claims(
+    already_satisfied_claims: QuerySet,
+    list_of_forced_payments: List[ForcedPaymentEvent],
+) -> QuerySet:
+    """
+    Removes duplicates from claims of settlement payments with pending transaction
+    which are on list of forced payments by adding exclusions to DepositClaim queryset.
+    """
+    for force_payment in list_of_forced_payments:
+        try:
+            matching_claim = already_satisfied_claims.get(tx_hash=force_payment.tx_hash)
```

Comment:
OK but please take into consideration that a trade off for this is that I have to re-search for the matching `force_payment` from `list_of_forced_payments` for each of the matching claims to compare . Still, should be faster than making many queries until ie. `list_of_forced_payments` is extremely large and it will have only matching element which will result in one query.

Reply:
I know. I think the queries are slowest operation here, not the Python loop.

Searching for each element by going over a list makes it O(n^2) but you could make it O(n log n) but having both lists sorted by `tx_hash` and then going over them in parallel in a single `while` loop. This would complicate the code a bit though so I think it's not worth it. Having a lot of forced payments between a single requestor-provider pair will be very rare. Each one requires a situation in which the requestor did not pay and the provider started UC5.